src/pycountry/__init__.py

```diff
@@ -126,101 +124,125 @@ class HistoricCountries(ExistingCountries):
 
 
-class LanguageFamilies(pycountry.db.Database):
+class LanguageFamily(pycountry.db.Data):
+    pass
+
+
+class LanguageFamilies(pycountry.db.Database[LanguageFamily]):
     """Provides access to an ISO 639-5 database
     (Language Families and Groups)."""
 
-    data_class = "LanguageFamily"
+    factory = LanguageFamily
     root_key = "639-5"
 
 
 class SubdivisionHierarchy(pycountry.db.Data):
-    def __init__(self, **kw):
-        if "parent" in kw:
-            kw["parent_code"] = kw["parent"]
-        else:
-            kw["parent_code"] = None
-        super().__init__(**kw)
-        self.country_code = self.code.split("-")[0]
-        if self.parent_code is not None:
-            # Split the parent_code to check if the country_code is already present
-            parts = self.parent_code.split("-")
-            if parts[0] != self.country_code:
-                self.parent_code = f"{self.country_code}-{self.parent_code}"
-
     @property
-    def country(self):
-        return countries.get(alpha_2=self.country_code)
+    def country(self) -> pycountry.db.Country:
+        return cast(
+            pycountry.db.Country, countries.get(alpha_2=self.country_code)
+        )
+
+    @cached_property
+    def country_code(self) -> str:
+        return self.code.split("-")[0]
 
     @property
-    def parent(self):
+    def parent(self) -> Optional["SubdivisionHierarchy"]:
         if not self.parent_code:
             return None
         return subdivisions.get(code=self.parent_code)
 
+    @cached_property
+    def parent_code(self) -> Optional[str]:
+        parent = self._fields.get("parent")
+        if parent is not None:
+            # check if the country_code is already present
+            parts = parent.split("-")
+            if parts[0] != self.country_code:
+                parent = f"{self.country_code}-{parent}"
+
+        return parent
+
 
-class Subdivisions(pycountry.db.Database):
+class Subdivisions(pycountry.db.Database[SubdivisionHierarchy]):
     # Note: subdivisions can be hierarchical to other subdivisions. The
     # parent_code attribute is related to other subdivisions, *not*
     # the country!
 
-    data_class = SubdivisionHierarchy
+    factory = SubdivisionHierarchy
     no_index = ["name", "parent_code", "parent", "type"]
     root_key = "3166-2"
 
-    def _load(self, *args, **kw):
-        super()._load(*args, **kw)
+    def _load(self) -> None:
+        super()._load()
 
         # Add index for the country code.
         self.indices["country_code"] = {}
         for subdivision in self:
-            divs = self.indices["country_code"].setdefault(
-                subdivision.country_code.lower(), set()
+            divs = cast(
+                Set[SubdivisionHierarchy],
+                self.indices["country_code"].setdefault(
+                    subdivision.country_code.lower(), set()  # type: ignore[arg-type]
+                ),
             )
             divs.add(subdivision)
```

Comment:
Is this actually correct? I typed the indices as `Dict[str, Dict[str, D]]`, but should they be instead typed as `Dict[str, Dict[str, Any]]`? I only quickly ran through the code and I'm not completely sure how or if this index is used. Widening the type would remove the need for the type ignore & cast, but then would allow much wider index types. Alternatively, this index could be declared separately so it could be typed more specifically.